test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl

```diff
@@ -0,0 +1,184 @@
+      a >>= 56;
+      a <<= 56;
+    }
+  } else if inputDataScheme == 3 {
+    // scheme 3: random ints, only a middle byte set
+    if describe then
+      writeln("generating random input setting only a middle byte");
+    fillRandom(array, seed=seed);
+    for a in array {
+      a >>= 56;
+      a <<= 56;
+      a >>= 16;
+    }
+  } else if inputDataScheme == 4 {
+    // scheme 4: random ints, only bottom byte set
+    if describe then
+      writeln("generating random input setting only the bottom byte");
+    fillRandom(array, seed=seed);
+    for a in array {
+      a &= 0xff;
+    }
+  } else if inputDataScheme == 5 {
+    // scheme 5: heavily skewed distribution,
+    // values are (1 << (random % 64))
+    if describe then
+      writeln("generating random powers of 2 input");
+    fillRandom(array, seed=seed);
+    for a in array {
+      var shift = mod(a, 64);
+      a = 1 << shift;
+    }
+  } else if inputDataScheme == 6 {
+    // scheme 6: data is already in sorted order
+    if describe then
+      writeln("generating already-sorted input");
+    array = 1..size;
+  } else if inputDataScheme == 7 {
+    // scheme 7: data is in reverse sorted order
+    if describe then
+      writeln("generating reverse-sorted input");
+    array = 1..size by -1;
+  }
+
+  return array;
+}
+
+proc testsize(size:int) {
+
+  var array = generateArray(size);
+
+  var nBytes = size*8;
+  var kibibytes = nBytes/1024.0;
+  var mibibytes = kibibytes/1024.0;
+  var sizestr = nBytes:string + " bytes";
+  if kibibytes >= 1.0 then
+    sizestr = kibibytes:string + " KiB";
+  if mibibytes >= 1.0 then
+    sizestr = mibibytes:string + " MiB";
+  if printStats then
+    writef("% 16s", sizestr);
+
+  var input = makeInput(array);
+
+  var ntrials = 1;
+  if printStats {
+    // use more trials for small problem sizes when measuring speed
+    if mibibytes < 1 then
+      ntrials = 100;
+    if kibibytes < 100 then
+      ntrials = 1_000;
+    if kibibytes < 0.4 then
+      ntrials = 10_000;
+  }
+
+  for m in methods {
+    var t: stopwatch;
+    for i in 1..ntrials {
+      input = makeInput(array);
+      on here.gpus[0]{
+        var arr = input; // Copy to gpu
+        t.start();
+        testsort(arr, m);
+        t.stop();
+        input = arr; // Copy back to cpu
+      }
+    }
+    var mibs = mibibytes * ntrials / t.elapsed();
+    // if printStats then writef(" % 12s sec  ", t.elapsed():string);
+    if printStats then writef(" % 12s MiB/s", mibs:string);
+
+    assert(isSorted(input));
+  }
+  if printStats then writeln();
+}
+
+proc main() {
+  // run generateArray to output the distribution
+  generateArray(100, describe=printStats);
+
+  if printStats {
+    writeln("Note, speeds are in MiB/s");
+    writef("% 16s", "size");
+  }
+
+  for m in methods {
+    if printStats then
+      writef(" % 16s", m);
+  }
+  if printStats then
+    writeln();
+
+  var size = minSize;
+  while size <= maxSize {
+    testsize(size);
+    size *= 2;
+  }
+}
```

Comment:
I haven't reviewed this file in detail. If you have parts that you need me to look at, I'd be happy to.

Reply:
I forgot to add a comment to the top of this file saying it's adapted from Michael's `test/library/packages/Sort/performance/sort-performance-explorer.chpl` 

I don't think it needs an in depth review :)